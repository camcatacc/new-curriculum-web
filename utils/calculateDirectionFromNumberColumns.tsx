// TODO: Simplify this function
function calculateDirectionFromNumberColumns(i: number, total: number, numPerRow: number): "fromLeft" | "fromRight" | "fromBottom" {
	// If everything does not fit in a row
	if (numPerRow < total) {
		// We calculate the number of items in the last row
		let nItemsLastRow = total % numPerRow;
		// We calculate the position from which the items will be in the last row
		let threshold = total - nItemsLastRow;
		// And we calculate the item position in that row
		let posInNewRow = i - numPerRow;

		// If the threshold is smaller than the item position and there are remaining items to be placed enter this conditional
		// If not, then the default method can be applied
		if (threshold < i && nItemsLastRow !== 0) {
			// If there is only one item in the last row, it must come from the bottom
			if (nItemsLastRow === 1) return "fromBottom";
			// We use the default method BUT with the position of the item IN THE LAST ROW
			switch (posInNewRow % nItemsLastRow) {
				case 0:
					return "fromRight";
				case 1:
					return "fromLeft";
				default:
					return "fromBottom";
			}
		}
	}

	switch (
		i % numPerRow //
	) {
		case 0:
			return "fromRight";
		case 1:
			return "fromLeft";
		default:
			return "fromBottom";
	}
}

export default calculateDirectionFromNumberColumns;
