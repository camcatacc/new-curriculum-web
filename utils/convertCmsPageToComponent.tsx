// Modules
import React from "react";

// Elements
import Abilities from "components/templates/Abilities/Abilities";
import Presentation from "components/templates/Presentation/Presentation";
import FindMe from "components/templates/FindMe/FindMe";

// Definitions
import type { ContentfulEntry } from "interfaces/cms/contentful";
import type { CmsPage } from "interfaces/cms/pages/Page";
import type { CmsPresentationPage } from "interfaces/cms/pages/Presentation";
import type { CmsAbilitiesPage } from "interfaces/cms/pages/Abilities";
import type { CmsFindMePage } from "interfaces/cms/pages/FindMe";

// Functions
const convertCmsPageToComponent = (page: ContentfulEntry<CmsPage>) => {
	switch (page.fields.type) {
		case "PresentationPage":
			return presentationCmsToComponent(page.fields.content.fields);
		case "AbilitiesPage":
			return abilitiesCmsToComponent(page.fields.content.fields);
		case "FindMePage":
			return findMeCmsToComponent(page.fields.content.fields);
	}
};

const presentationCmsToComponent = (page: CmsPresentationPage) => <Presentation {...page} />;
const abilitiesCmsToComponent = (page: CmsAbilitiesPage) => {
	const lists = page.abilitiesLists.map((abList) => {
		return {
			...abList.fields
		};
	});

	return <Abilities title={page.title} abilitiesLists={lists} />;
};

const findMeCmsToComponent = (page: CmsFindMePage) => <FindMe {...page} />;

export default convertCmsPageToComponent;
