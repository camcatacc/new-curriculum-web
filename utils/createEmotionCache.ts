// Modules
import createCache from "@emotion/cache";

// Functions
export default function createEmotionCache() {
	return createCache({ key: "css" });
}
