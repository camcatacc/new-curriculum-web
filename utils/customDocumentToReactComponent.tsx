// Modules
import React from "react";
import { BLOCKS, MARKS } from "@contentful/rich-text-types";
import { documentToReactComponents } from "@contentful/rich-text-react-renderer";

// Styles
import { secondary } from "styles/customColors";

// Elements
import { Paragraph } from "@hardru/curriculum-web-components";

// Definitions
import type { Document } from "@contentful/rich-text-types";
import type { Options } from "@contentful/rich-text-react-renderer";
import type { PropTypes, Theme, TypographyVariant } from "@material-ui/core";

export interface CustomDocumentToReactComponentOptions {
	theme: Theme;
	boldColor?: boolean;
	marginBottom?: 1 | 2 | 3 | 4 | 6 | 8 | 10;
	align?: PropTypes.Alignment;
	variant?: TypographyVariant;
}

// Function
function customDocumentToReactComponent(document: Document, { boldColor, align, variant, theme }: CustomDocumentToReactComponentOptions) {
	let options: Options = {
		renderNode: {
			[BLOCKS.PARAGRAPH]: (_: any, children: any) => (
				<div>
					<Paragraph variant={variant} align={align}>
						{children}
					</Paragraph>
				</div>
			)
		}
	};

	if (boldColor) {
		options = { ...options, renderMark: {} };
		options.renderMark![MARKS.BOLD] = (text) => <b style={{ color: theme.palette.secondary[theme.palette.type] }}>{text}</b>;
	}

	return documentToReactComponents(document, options);
}

export default customDocumentToReactComponent;
