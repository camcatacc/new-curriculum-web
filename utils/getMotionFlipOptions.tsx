// Definitions
import type { MotionProps, Variants } from "framer-motion";

export interface GetMotionFlipOptionsOptions {
	amount?: number;
	delay?: number;
}

// Functions
function getMotionFlipOptions(animated: boolean, { amount, delay }: GetMotionFlipOptionsOptions): MotionProps {
	if (!animated) return {};
	const variants = getMotionFlipVariants(animated, delay);
	return {
		variants: variants,
		viewport: {
			amount,
			once: true
		},
		initial: "offscreen",
		whileInView: "onscreen"
	};
}

export function getMotionFlipVariants(animated: boolean, delay?: number): Variants {
	if (!animated) return {};
	return {
		offscreen: {
			rotateY: 90,
			opacity: 0
		},
		onscreen: { rotateY: 0, opacity: 1, transition: { duration: 0.5, delay: delay } }
	};
}

export default getMotionFlipOptions;
