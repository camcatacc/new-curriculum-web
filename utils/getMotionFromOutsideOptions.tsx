// Definitions
import type { MotionProps, Variants } from "framer-motion";

export type AnimatedDirection = "fromLeft" | "fromRight" | "fromBottom";

// Functions
export function getOnScrollMotionFromOutsideOptions(animated: boolean, animatedDirection: AnimatedDirection, delay?: number): MotionProps {
	if (!animated) return {};

	const { initial, animation } = getMotionFromOutsideVariants(animatedDirection, delay);

	const variants: Variants = {
		offscreen: initial,
		onscreen: animation
	};

	return {
		variants,
		initial: "offscreen",
		whileInView: "onscreen",
		viewport: { once: true, amount: 0.6 }
	};
}

export function getMotionFromOutsideVariants(animatedDirection: AnimatedDirection, delay?: number): Variants {
	let initialX = 0,
		initialY = 0;

	switch (animatedDirection) {
		case "fromLeft":
			initialX = -50;
			break;
		case "fromRight":
			initialX = 50;
			break;
		case "fromBottom":
			initialY = 50;
			break;
	}
	return {
		initial: { x: initialX, y: initialY, opacity: 0 },
		animation: {
			x: 0,
			y: 0,
			opacity: 1,
			transition: {
				type: "tween",
				duration: 1,
				delay: delay,
				when: "beforeChildren"
			}
		}
	};
}

export default getOnScrollMotionFromOutsideOptions;
