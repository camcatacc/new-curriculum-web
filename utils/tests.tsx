// Modules
import React, { ReactElement } from "react";
import { ThemeProvider } from "@material-ui/styles";

// Testing modules
import { render, RenderOptions, RenderResult } from "@testing-library/react";

// Redux Modules
import { configureStore } from "@reduxjs/toolkit";

// Redux Components
import rootReducer from "redux/rootReducer";
import { Provider } from "react-redux";
import { defaultTestRootState } from "redux/store.test";

// Auxiliary functions
import { getMaterialTheme } from "styles/materialUi";

// Definitions
import type { RootState } from "redux/types";

// Functions
export const customRender = (ui: ReactElement, state: Partial<RootState> = defaultTestRootState, renderOptions?: RenderOptions & { darkMode?: boolean }): RenderResult => {
	const Wrapper: React.FC = ({ children }) => {
		const store = configureStore({
			reducer: rootReducer,
			preloadedState: state
		});

		return (
			<ThemeProvider theme={getMaterialTheme(renderOptions?.darkMode ? "dark" : "light")}>
				<Provider store={store}>{children}</Provider>
			</ThemeProvider>
		);
	};

	return render(ui, { wrapper: Wrapper, ...renderOptions });
};
