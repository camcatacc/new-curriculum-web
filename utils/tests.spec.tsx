import { customRender } from "utils/tests";
describe("customRender snapshots", () => {
	it("customRender with light mode", () => {
		const { container } = customRender(<p>Custom Render</p>);
		expect(container).toMatchSnapshot();
	});
	it("customRender with dark mode", () => {
		const { container } = customRender(<p>Custom Render</p>, undefined, { darkMode: true });
		expect(container).toMatchSnapshot();
	});
});
