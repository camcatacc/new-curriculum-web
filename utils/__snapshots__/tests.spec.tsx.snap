// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customRender snapshots customRender with dark mode 1`] = `
<div>
  <p>
    Custom Render
  </p>
</div>
`;

exports[`customRender snapshots customRender with light mode 1`] = `
<div>
  <p>
    Custom Render
  </p>
</div>
`;
